8302818: Optimize wrapper sets and immutable sets of Enums

yuantj opened the pull request:
Currently, the two subclasses of `java.util.EnumSet` optimize bulk operations when the argument is also a `EnumSet`, but there is no such optimization for wrapper sets (returned by `Collections.unmodifiableSet`, `Collections.synchronizedSet`, etc.) and immutable sets (returned by `Set.of` methods) of `Enum`s.

This PR introduces optimization classes for these situations.  No public APIs are changed.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302818](https://bugs.openjdk.org/browse/JDK-8302818): Optimize wrapper sets and immutable sets of Enums (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12498/head:pull/12498` \
`$ git checkout pull/12498`

Update a local copy of the PR: \
`$ git checkout pull/12498` \
`$ git pull https://git.openjdk.org/jdk.git pull/12498/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12498`

View PR using the GUI difftool: \
`$ git pr show -t 12498`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12498.diff">https://git.openjdk.org/jdk/pull/12498.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12498#issuecomment-1436901865)

liach commented:
Just curious, how does this enum set change affect the performance of construction of regular sets via Set.of calls?

liach commented:
On the JBS there's https://bugs.openjdk.org/browse/JDK-8145048 which appears to fit your changes.

yuantj commented:
> Just curious, how does this enum set change affect the performance of construction of regular sets via Set.of calls?

See `ImmutableCollections.setFromArray`, it relies on the fact that `EnumSet.copyOf` will throw a `ClassCastException` if the elements are not enum constants of a same enum class.  Trying to catch an exception is a bit expensive (so maybe I can optimize that).

yuantj commented:
> On the JBS there's https://bugs.openjdk.org/browse/JDK-8145048 which appears to fit your changes.

I only enhanced `EnumSet`, but not `EnumMap` (yet) so I don't think it fits my changes.  Maybe I have to submit an enhancement with [the webbug system](https://bugreport.java.com/bugreport/).

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -890,6 +890,39 @@ public <T> T[] toArray(T[] a) {
+    @SafeVarargs
+    @SuppressWarnings({"varargs", "unchecked", "rawtypes"})
+    static <E> Set<E> setFromArray(E... input) {
+        if (input.length == 0) {
```
This special case can be removed, as no caller calls with an array with length <= 2

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -890,6 +890,39 @@ public <T> T[] toArray(T[] a) {
+        EnumSet<?> es;
+        try {
+            // Sorry, I have to use a raw type here.
+            es = EnumSet.copyOf((Collection)Arrays.asList(input));  // rejects null
```
This fails to throw `IllegalArgumentException` on duplicated elements.

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1095,305 @@ public <T> T[] toArray(T[] a) {
+            long unseen = elements;
+            long lastReturned = 0;
+            final Enum<?>[] universe
+                = SharedSecrets.getJavaLangAccess().getEnumConstantsShared(elementType);
```
I think the convention is to put `SharedSecrets.getJavaLangAccess()` instance in a static final field like `JLA` than calling it every time

liach commented:
Just took a peek again. Besides the critical error of not throwing IAE for duplicate inputs (which you should probably add new test cases for), I think the 2 implementations share the same `toArray` `hashCode`, and the `contains` logic can be pulled to call an abstract `contains(int ordinal)` instead.

liach commented on src/java.base/share/classes/java/util/Collections.java:
```diff
@@ -1164,6 +1170,81 @@ public static <T> Set<T> unmodifiableSet(Set<? extends T> s) {
+        @java.io.Serial
+        private static final long serialVersionUID = -1110577510253015312L;
+
+        final RegularEnumSetCompatible<? extends E> es;
```
Can't you just use `<E>` as the type argument instead, for all occurrences? Can avoid a cast in `elementType()` too, and it's compatible argument to superclass constructor.

liach commented on src/java.base/share/classes/java/util/Collections.java:
```diff
@@ -1150,6 +1150,12 @@ public static <T> Set<T> unmodifiableSet(Set<? extends T> s) {
         if (s.getClass() == UnmodifiableSet.class) {
             return (Set<T>) s;
         }
+        if (s instanceof RegularEnumSetCompatible) {
```
Why not use `if (s instanceof RegularEnumSetCompatible es)` instead?

liach commented on src/java.base/share/classes/java/util/Collections.java:
```diff
@@ -2228,6 +2323,97 @@ public boolean equals(Object o) {
+        @Override
+        public long[] elements() {
+            synchronized (mutex) {
+                return es.elements();
```
Is this really safe? `elements()` in `JumboEnumSet` returns the backing array, so this returned array is mutable. Should perform `es.elements().clone()` for more safety.

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1095,305 @@ public <T> T[] toArray(T[] a) {
+
+                return (es.elements() & ~elements) == 0;
+            } else {
+                for (Object o : c) {
```
Can just call super here

liach commented on src/java.base/share/classes/java/util/JumboEnumSetCompatible.java:
```diff
@@ -0,0 +1,39 @@
+                Collections.SynchronizedJumboEnumSet,
+                ImmutableCollections.ImmutableJumboEnumSet {
+    Class<E> elementType();
+    long[] elements();
```
Since long arrays are mutable (unlike the `long elements()` in regular set), might add a note about this, as enum sets never throws CMEs

yuantj commented:
@liach Thanks for reviewing. I've fixed the issues you've mentioned.

liach commented:
Have you submitted an issue at https://bugreport.java.com/bugreport/ ? It will show up a few days later on the JBS. If not, I can create an issue on the JBS for you too as long as you provide a prompt.

yuantj commented:
> Have you submitted an issue at https://bugreport.java.com/bugreport/ ? It will show up a few days later on the JBS. If not, I can create an issue on the JBS for you too as long as you provide a prompt.

No I haven't. Submit one for me please. Thanks!

liach commented:
Created as https://bugs.openjdk.org/browse/JDK-8302818

I strongly recommend you to add tests of pure-enum and mixed-enum immutable set creation to ensure your code is logically correct. You seems to be developing the JDK in some weird way as you upload changes directly to GitHub Web UI.
I recommend developing according to the docs: see [IDE](https://github.com/openjdk/jdk/blob/master/doc/ide.md) and [building](https://github.com/openjdk/jdk/blob/master/doc/building.md) docs.

Also, try set your console's code page to 437 (US-ASCII), as some non-English consoles have caused build failures for me before. If you need help, you can leave this PR as a work in progress (so we don't spam the openjdk mailing lists) and I can answer your questions about the setup.

ExE-Boss commented on src/java.base/share/classes/java/util/Collections.java:
```diff
@@ -1150,6 +1150,12 @@ public static <T> Set<T> unmodifiableSet(Set<? extends T> s) {
         if (s.getClass() == UnmodifiableSet.class) {
             return (Set<T>) s;
         }
```
To avoid re‑wrapping already unmodifiable `EnumSet`s, this should also check for those:
```suggestion
        if (s.getClass() == UnmodifiableSet.class
            || s.getClass() == UnmodifiableRegularEnumSet.class
            || s.getClass() == UnmodifiableJumboEnumSet.class) {
            return (Set<T>) s;
        }
```

yuantj commented:
Oops, my bad. Please ignore the merge warning.

I've just add several tests for `Set.of` factories.

yuantj commented:
By the way, I've just built the JDK successfully on my Mac, and the tests run successfully too. Let me know if it fails on other operating systems.

liach commented on src/java.base/share/classes/java/util/Set.java:
```diff
@@ -728,10 +728,25 @@ static <E> Set<E> of(E... elements) {
+        }
+        E e1 = it.next();
+        if (!it.hasNext()) {
+            return Set.of(e0, e1);
```
Bad change, doesn't handle e0.equals(e1), and this is getting beyond the original issue

yuantj replied:
> Bad change, doesn't handle e0.equals(e1), and this is getting beyond the original issue

`Set.of(E, E)` handles duplication. If `coll` is `{Jumbo,Regular}EnumSetCompatible`, then we don't have to create the `HashSet`.

liach replied:
no, Set.copyOf explicitly allows duplications, unlike Set.of. You need to read the specs so you know what you are doing.

liach commented on src/java.base/share/classes/java/util/EnumSet.java:
```diff
@@ -178,8 +179,12 @@ public static <E extends Enum<E>> EnumSet<E> copyOf(Collection<E> c) {
+                result.addAll(c);   // optimized for compatible sets
+            } else {
+                while (i.hasNext())
+                    result.add(i.next());
```
Can't we just use addAll in all cases?

yuantj replied:
I wonder why the original code didn't use `addAll`, so I didn't change it. 😂

yuantj replied:
> I wonder why the original code didn't use `addAll`, so I didn't change it. 😂

It seems to be okay to use `addAll` in all cases. Do we need extra test cases for this?

cl4es commented:
I'm wary of the impact of adding new wrapper classes. It may make the factory methods slightly slower due additional checks, but also risks increasing the number of classes at various call-sites which might upset call site inlining. 

An alternative design which would avoid adding more classes could be to add package-private accessors to the existing `Unmodifiable/Synchronized` wrapper classes so that `EnumSet/-Map` can retrieve the underlying set of an unmodifiable or synchronized `Set` or `Map` and then use it directly for these bulk operations. Then you'd contain the additional overhead to `EnumSet/-Map`.

yuantj commented:
Thanks again for reviewing. Correct me if I'm wrong.

@cl4es I need to add these new wrapper classes `{Unmodifiable,Synchronized}{Regular,Jumbo}EnumSet` to implement interfaces `{Regular,Jumbo}EnumSetCompatible`, just like `*RandomAccessList` which appear to implement `RandomAccess`. I don't think removing these wrapper classes and introducing accessors is a good idea, because I will have to check for the wrapper classes in bulk operation methods, which are used more frequently.

@pavelrappo `{Unmodifiable,Sychronized}Set` also have subclasses implementing `SortedSet` and `NavigableSet`, I'm not sure whether `instanceof` is safe for these subclasses.

liach commented:
The `RandomAccess` implementation was public to user while the wrapper classes aren't. So I guess we can implement another interface for these wrappers to retrieve their backing instances, like:
```java
interface WrapperCollection {
  Collection<?> getBacking();
}
```
and thus:
```java
interface RegularEnumSetCompatible {
  static RegularEnumSetCompatible tryConvert(Collection<?> coll) {
    if (coll instanceof RegularEnumSetCompatible compat) return compat;
    if (coll instanceof WrapperCollection wrap) return tryConvert(wrap);
    return null;
  }
}
```

Adding extra `getClass() == UnmodifiableRegularEnumSet.class` indeed will affect most users, which rarely are enum set compatibles.

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1102,266 @@ public <T> T[] toArray(T[] a) {
+        @Stable
+        final long[] elements;
+
+        @Stable
```
Useless

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1102,266 @@ public <T> T[] toArray(T[] a) {
+            implements Serializable permits ImmutableRegularEnumSet, ImmutableJumboEnumSet {
+        static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+        @Stable
```
Useless

liach commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1102,266 @@ public <T> T[] toArray(T[] a) {
+    @jdk.internal.ValueBased
+    static final class ImmutableRegularEnumSet<E extends Enum<E>> extends AbstractImmutableEnumSet<E>
+            implements RegularEnumSetCompatible<E>, Serializable {
+        @Stable
```
Useless

liach reviewed (commented):
Also, in JEP 441, there is pattern matching support for different enums implementing one interface in one switch, see the enum constants section there. I wonder how useful this is, and how often do we have interfaces with more than one enum implementations (e.g. StandardOptions/ExtendedOptions)

cl4es commented:
If this level of complexity is indeed needed to get whatever improvement you're after then I don't see how this can be worth its weight. Microbenchmarking might help support your case here, but assessing the potential performance costs from gradually increasing the number of classes floating around at various call sites in arbitrary applications can be tough and is something we need to be very careful not to do without solid evidence.

yuantj commented:
I have deleted `{Regular,Jumbo}EnumSetCompatible` interfaces and introduced some package-private methods in `AbstractCollection`. This avoids rarely-used checks from static factories in `Collections` and does not reduce much performance compared against the previous implementation.

ExE-Boss commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1102,275 @@ public <T> T[] toArray(T[] a) {
+
+        private final class ImmutableRESIterator implements Iterator<E> {
+            long unseen = elements;
+            long lastReturned = 0;
```
```suggestion
```

ExE-Boss commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1102,275 @@ public <T> T[] toArray(T[] a) {
+            public E next() {
+                if (unseen == 0)
+                    throw new NoSuchElementException();
+                lastReturned = unseen & -unseen;
```
```suggestion
                long lastReturned = unseen & -unseen;
```

ExE-Boss commented on src/java.base/share/classes/java/util/ImmutableCollections.java:
```diff
@@ -1062,6 +1102,275 @@ public <T> T[] toArray(T[] a) {
+        private final class ImmutableJESIterator implements Iterator<E> {
+            long unseen = elements[0];
+            int unseenIndex = 0;
+            long lastReturned = 0;
+            int lastReturnedIndex = 0;
```
```suggestion
```

ExE-Boss reviewed (commented):
The `Immutable*ESIterator.lastReturned*` fields can be made into locals of `Immutable*ESIterator::next()`, as the only reason those fields exist in `RegularEnumSet.EnumSetIterator` and `JumboEnumSet.EnumSetIterator` is to support `Iterator::remove()`, which is not a valid operation on immutable enum sets:

stuart-marks commented:
I agree with Claes here in my skepticism of the usefulness of this sort of change. It seems like it's adding lots of complexity. In addition, it also seems like there's a lot of churn in the design. I'd suggest returning this PR to draft status until it can be determined that there's a clear, net improvement.

yuantj commented:
@cl4es @stuart-marks Thanks for reviewing and commenting. I'm converting this PR to draft until I finish evaluating whether these changes are necessary or not.
